### ISSUE TITLE
gzweb7 Docker build fails

### ISSUE THREAD

**Issue by damonkohler** [osrf/docker_images#43]
I'm currently trying to build the [gzweb7 Docker image](https://github.com/osrf/docker_images/blob/master/gazebo/gazebo7/gzweb7/Dockerfile) and running into an issue generating thumbnails.

First, I had to change the package version numbers from 7.4 to 7.5 [here](https://github.com/osrf/docker_images/blob/master/gazebo/gazebo7/gzserver7/Dockerfile#L15), [here](https://github.com/osrf/docker_images/blob/master/gazebo/gazebo7/gzweb7/Dockerfile#L26), and [here](https://github.com/osrf/docker_images/blob/master/gazebo/gazebo7/libgazebo7/Dockerfile#L8).

After that, building gzweb7 succeeds (exit code 0) but fails to generate thumbnails with the same errors over and over:

```
Creating thumbnail for walkway_metal_straight
gzserver: /usr/include/boost/smart_ptr/shared_ptr.hpp:653: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::rendering::Scene; typename boost::detail::sp_member_access<T>::type = gazebo::rendering::Scene*]: Assertion `px !=
 0' failed.
./tools/gzthumbnails.sh: line 6: 18811 Aborted                 (core dumped) gzserver -s libModelPropShop.so $DIR/green.world --propshop-save "$ASSETS/${dir##*/}/thumbnails" --propshop-model "$ASSETS/${dir##*/}/model.sdf"
convert.im6: unable to open image `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/1.png': No such file or directory @ error/blob.c/OpenBlob/2641.
convert.im6: unable to open file `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/1.png' @ error/png.c/ReadPNGImage/3672.
convert.im6: no images defined `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png' @ error/convert.c/ConvertImageCommand/3044.
convert.im6: unable to open image `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png': No such file or directory @ error/blob.c/OpenBlob/2641.
convert.im6: unable to open file `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png' @ error/png.c/ReadPNGImage/3672.
convert.im6: no images defined `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png' @ error/convert.c/ConvertImageCommand/3044.
convert.im6: unable to open image `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png': No such file or directory @ error/blob.c/OpenBlob/2641.
convert.im6: unable to open file `/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/0.png' @ error/png.c/ReadPNGImage/3672.
convert.im6: image sequence is required `+clone' @ error/convert.c/ConvertImageCommand/933.
rm: cannot remove '/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/1.png': No such file or directory
rm: cannot remove '/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/2.png': No such file or directory
rm: cannot remove '/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/3.png': No such file or directory
rm: cannot remove '/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/4.png': No such file or directory
rm: cannot remove '/root/gzweb/tools/../http/client/assets/walkway_metal_straight/thumbnails/5.png': No such file or directory
```

**Comment by ruffsl** [comment]
Yes, I had this issue when I first created the Dockerfile. It trouble is that gazebo still needs an X Server to render scenes. I opened an issue related to this on Gazebo's issue tracker, see:
https://bitbucket.org/osrf/gazebo/issues/1602/removing-x11-dependency-for-official

You can get around this by running the container from the resulting image with missing thumbnails and starting it with an attached X server:
http://wiki.ros.org/docker/Tutorials/GUI

Then rerun the build command, and then commit that connter as a new image layer so you can use the fix image that now thumbnails in the Web GUI:
https://docs.docker.com/engine/reference/commandline/commit/

Obviously, fixing the build to work without an X server would be way more preferable. Perhaps @scpeters , @j-rivero or @nkoenig  would have better ideas, like downloading pre-rendered thumbnails that are provided alongside each 3D model?

**Comment by damonkohler** [comment]
I managed to get it to work by using xvfb. It's quite slow, though. I wasn't able to get it working using nvidia-docker, but perhaps more persistence would have seen me through.

**Comment by ruffsl** [comment]
@damonkohler , could you share your solution using xvfb? 
I'd be interested in seeing how it's used, and if nvidia-docker could be worked in.

**Comment by scpeters** [comment]
@damonkohler I have some instructions for setting up nvidia-docker with a gazebo8 dockerfile over here:

* https://bitbucket.org/scpeters/unix-stuff/src/4a1228a910c29dd89306b30466c1b6742a4b5b6b/gazebo8-docker/?at=master

There's a few tricky bits, and one point at which I don't know how to get around restarting the system, but several people have followed those instructions successfully now. Maybe it will work for gzweb too?

**Comment by damonkohler** [comment]
Sorry for the slow reply. To use xvfb, I simply executed deploy.sh like so:

`xvfb-run -s "-screen 0 1280x1024x24" ./deploy.sh`

Using that, I was able to build gzweb (from the default branch) in a Docker container.

My original goal was to get gzweb working with a Freight simulation. Unfortunately, I wasn't able to get the Freight mesh displaying in gzweb. I tried building it like so:

```
export GAZEBO_MODEL_PATH=/root/catkin_ws/install/share:/root/catkin_ws/install/share/fetch_gazebo/models:${GAZEBO_MODEL_PATH}
xvfb-run -s "-screen 0 1280x1024x24" ./deploy.sh -m local
```

I also tweaked the launch file to include -package_to_model when running the URDF spawner. Not only did that not result in the mesh being displayed, but it also put the Freight into free fall in the simulator (i.e. it fell through the ground plane).

I've had to move on to other tasks for now. But, I'm happy to try out suggested fixes to these issues!
